File: classic/Code/BinaryTreeLevelOrderTraversal.cpp
/**

 * Definition for binary tree

 * struct TreeNode {

 *     int val;

 *     TreeNode *left;

 *     TreeNode *right;

 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}

 * };

 */

class Solution {

public:

    vector<vector<int> > levelOrder(TreeNode *root) {

        vector<vector<int> > r;

        if (root==NULL) return r;

        queue<TreeNode*>q;

        q.push(root);

        int c=1,n=0,v=0;

        vector<int>l;

        while(q.size()!=0){

            TreeNode * no=q.front();

            q.pop();

            v++;

            l.push_back(no->val);

            if(no->left!=NULL){

                q.push(no->left);

                n++;

            }

            if(no->right!=NULL){

                q.push(no->right);

                n++;

            }

            if(v==c){

                v=0;

                c=n;

                n=0;

                r.push_back(l);

                l.clear();

            }

        }

        return r;

    }

};

File: classic/Code/BinaryTreeLevelOrderTraversal.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer> >result=new ArrayList<List<Integer> >();
        if(root==null){
            return result;
        }
        Queue<TreeNode>current_level=new LinkedList<TreeNode>();
        Queue<TreeNode>next_level=new LinkedList<TreeNode>();
        current_level.add(root);
        while(!current_level.isEmpty()){
            List<Integer>currentResult=new ArrayList<Integer>();
            while(!current_level.isEmpty()){
                TreeNode node=current_level.poll();
                currentResult.add(node.val);
                if(node.left!=null){
                    next_level.add(node.left);
                }
                if(node.right!=null){
                    next_level.add(node.right);
                }
            }
            result.add(currentResult);
            Queue<TreeNode>tmp=current_level;
            current_level=next_level;
            next_level=tmp;
        }
        return result;
    }
}
